pkg/types/automation/trigger.go

```diff
@@ -0,0 +1,87 @@
+package automation
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+)
+
+// Trigger represents a trigger for an upkeep.
+// It contains an extension per trigger type, and the block number + hash
+// in which the trigger was checked.
+// NOTE: This struct is sent on the p2p network as part of observations to get quorum
+// Any change here should be backwards compatible and should keep validation and
+// quorum requirements in mind. Please ensure to get a proper review along with an
+// upgrade plan before changing this
+type Trigger struct {
+	// BlockNumber is the block number in which the trigger was checked
+	BlockNumber BlockNumber
+	// BlockHash is the block hash in which the trigger was checked
+	BlockHash [32]byte
+	// LogTriggerExtension is the extension for log triggers
+	LogTriggerExtension *LogTriggerExtension
+}
+
+func (r Trigger) String() string {
+	res := fmt.Sprintf(`{"BlockNumber":%d,"BlockHash":"%s"`, r.BlockNumber, hex.EncodeToString(r.BlockHash[:]))
+	if r.LogTriggerExtension != nil {
+		res += fmt.Sprintf(`,"LogTriggerExtension":%s`, r.LogTriggerExtension)
+	}
+	res += "}"
+	return res
+}
+
+// NewTrigger returns a new basic trigger w/o extension
+func NewTrigger(blockNumber BlockNumber, blockHash [32]byte) Trigger {
+	return Trigger{
+		BlockNumber: blockNumber,
+		BlockHash:   blockHash,
+	}
+}
+
+func NewLogTrigger(blockNumber BlockNumber, blockHash [32]byte, logTriggerExtension *LogTriggerExtension) Trigger {
+	return Trigger{
+		BlockNumber:         blockNumber,
+		BlockHash:           blockHash,
+		LogTriggerExtension: logTriggerExtension,
+	}
+}
+
+// LogTriggerExtension is the extension used for log triggers,
+// It contains information of the log event that was triggered.
+// NOTE: This struct is sent on the p2p network as part of observations to get quorum
+// Any change here should be backwards compatible and should keep validation and
+// quorum requirements in mind. Please ensure to get a proper review along with an
+// upgrade plan before changing this
+type LogTriggerExtension struct {
+	// LogTxHash is the transaction hash of the log event
+	TxHash [32]byte
```

Comment:
I see a lot of `[32]byte` hash types. Is that something we will have to relax later on for other chains? like to a slice `[]byte`?

How are these values actually used? The approach taken with accounts is to just use a `string`, which for EVM happens to be the encoded hex format, but then never try to parse those values - only compare them directly. It would be nice to be able to make a similar kind of simplification here.

Reply:
The TxHash is used to build a log identifier for a log event; I think we use the block hash as well and a few other fields maybe? I think we also hex encode this field for a string representation as well!

Reply:
Having these as concrete `[32]bytes` gives us a lot of safety as these values are sent within observations and outcomes in OCR, and having a tight control on the size reduces surface area for potential security issues and gives us better control over how many upkeeps we can include in an observation/outcome. 

We haven't run into chains where this has been an issue yet, but if we generalise this to byte slice then will need to do a proper review of any downstream implications.